File: 1036-rotting-oranges/rotting-oranges.py
from collections import deque
from typing import List

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:

        number_of_oranges = 0
        m, n = len(grid), len(grid[0])
        queue = deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    queue.append((i, j))
                if grid[i][j] == 1:
                    number_of_oranges += 1

        if number_of_oranges == 0:
            return 0
        
        dir = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        total_time = -1

        while queue:
            total_time += 1
            for _ in range(len(queue)):
                x, y = queue.popleft()

                for dx, dy in dir:
                    nx, ny = dx + x, dy + y
                    if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                        grid[nx][ny] = 2
                        number_of_oranges -= 1
                        queue.append((nx, ny))

        return total_time if number_of_oranges == 0 else -1
        
        

        

        queue = deque([start])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        while queue:
            x, y = queue.popleft()

            for dx, dy in directions:
                nx, ny = dx + x, dy + y
                if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                    grid[nx][ny] = 2
                    number_of_oranges -= 1
                    queue.append((nx, ny))


        print("number of oranges after : ", number_of_oranges)
        print("total time taken : ", total_time)

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {


        int fresh = 0;
        Queue<int[]> rotten = new LinkedList<>();

        int m = grid.length;
        int n = grid[0].length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2){
                    rotten.offer(new int[] {i, j});
                }
                if (grid[i][j] == 1) {
                    fresh++;
                }
            }
        }
        System.out.println("fresh -> " + fresh);

        if (fresh == 0){
            return 0;
        }

        int minutes = 0;
        int[][] directions = new int[][] {{0,1}, {1,0}, {-1,0}, {0,-1}};
        
        while (!rotten.isEmpty()){
            boolean rotted = false;
            int rottenQueueSize = rotten.size();

            for (int i = 0; i < rottenQueueSize; i++) {
                int[] loc = rotten.poll();
                int x = loc[0];
                int y = loc[1];

                for (int[] dir : directions){
                    int nx = x + dir[0];
                    int ny = y + dir[1];

                    if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == 1){
                        grid[nx][ny] = 2;
                        rotten.offer(new int[] {nx, ny});
                        fresh--;
                        rotted = true;
                    }
                }
            }
            if (rotted) {
                minutes++;
            }

        }
        return fresh == 0 ? minutes : -1;
    }
}